README
Gnome2::GConf version 1.043
===========================

This module allows you to use the GConf configuration system in order to
store/retrieve the configuration of an application.  GConf is a powerful
configuration manager based on a user daemon that handles a set of key and
value pairs, and notifies any changes of the value to every program that
monitors those keys.  GConf is used by GNOME 2.x.


INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install


To avoid installing to a system directory, since this is a beta release,
you can change the installation prefix at Makefile.PL time with

   perl Makefile.PL PREFIX=/some/other/place

This will install the module to the subdirectory lib/perl5 under the given
prefix.  If this is not already in perl's include path, you'll need to tell
perl how to get to this library directory so you can use it; there are three
ways:

  in your environment (the easiest):
    PERL5LIB=/some/other/place/lib/perl5/site_perl:$PERL5LIB

  on the perl command line:
    perl -I /some/other/place/lib/perl5/site_perl yourscript

  in the code of your perl script:
    use lib '/some/other/place/lib/perl5/site_perl';
  

DEPENDENCIES

This module requires these other modules and libraries:

  perl >= 5.8.0
  Glib >= 1.120 (perl module and requisite C libraries)
  libgconf2 >= 2.0.0 (C library)

To build from source, you also need:

  ExtUtils::Depends >= 0.2
  ExtUtils::PkgConfig >= 1.03

COPYRIGHT AND LICENSE

Copyright (C) 2003-2006 by Emmanuele Bassi <emmanuele.bassi@iol.it>

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Library General Public
License as published by the Free Software Foundation; either
version 2 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Library General Public License for more details.

You should have received a copy of the GNU Library General Public
License along with this library; if not, write to the 
Free Software Foundation, Inc., 59 Temple Place - Suite 330, 
Boston, MA  02111-1307  USA.


NEWS
Overview of changes since 1.042
===============================

* Treat Gnome2::GConf::Schema objects as fundamental types
  in Gnome2::GConf::Value.
* Fix usage of Gnome2::GConf::Client getters for keys of type
  'schema' (#361103, Laurent Simonneau)
* Fix a leak inside Gnome2::GConf::Schema.

Overview of changes since 1.041
===============================

* Implement GConf::Client::recursive_unset() to recursively
  unset a directory in the database. (#361094, based on a patch
  by Laurent Simonneau)

Overview of changes since 1.040
===============================

* Fix a leak in GConfValue. (#358294, Santeri Paavolainen)

Overview of changes since 1.031
===============================

* Fix loaded flags on Darwin. [Daniel Macks]

Overview of changes since 1.020
===============================

* The "An Address Indicates Where It Is" release.
* Wrap Gnome2::GConf::Engine::get_for_addresses. [Emmanuele]

Overview of changes since 1.013
===============================

* The "Differential Engine" release
* Fix Gnome2::GConf::Engine reference mechanism. [Emmanuele] 
* Use Glib::CodeGen from the stable Glib release. [Emmanuele]

Overview of changes since 1.012
===============================

* The "Breaking up is hard to do" release.
* Implemented the new error handling system. [Emmanuele]
* Changed Gnome2::GConf::Client::all_entries semantics. [Emmanuele]
* Documentation fixes. [Emmanuele]

Overview of changes since 1.011
===============================

* Finished the GConfEngine bindings. [Emmanuele]

Overview of changes since 1.010
===============================

* Require GConf 2.0.0. [Emmanuele]

Overview of changes since 1.000
===============================

* Wrapped GConfEngine. [Emmanuele]
* Documentation fixes. [Emmanuele]
* Require Glib 1.070. [Emmanuele]

Overview of changes since 0.94
==============================

* New test suite for GConfClient. [Emmanuele]
* Adapted version information methods to the current policy. [Toersten]
* Updated documentation. [Emmanuele]
* Require Glib 1.040. [Emmanuele]

Overview of changes since 0.93
==============================

* Fixed a stupid bug in GConfValue [Emmanuele]

Overview of changes since 0.92

* Split-up API documentation for GConfValue, GConfEntry,
  GConfChangeSet and GConfSchema. [Emmanuele]
* API frozen [Emmanuele]
* Require Glib 1.038 (for API freeze state) [Emmanuele]

Overview of changes since 0.91
==============================

* Require Glib 1.037 [Emmanuele]
* Use the new Glib::Error exception handling. [Emmanuele]
* Updated documentation. [muppet]

Overview of changes since 0.46
==============================

* Require ExtUtils::PkgConfig 1.03
* Require Glib 1.020
* Remove runtime dependency on Gtk2; Gtk2's Gtk2::CodeGen is a
  build-only requirement.  [muppet]
* Build system updates.  [muppet]
* Updated documentation.  [Emmanuele]
* Added some test methods that apps should not use.  [Emmanuele]

[started NEWS]


GConf.pm
#
# Copyright (c) 2003-2006 by Emmanuele Bassi (see the file AUTHORS)
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Library General Public
# License as published by the Free Software Foundation; either
# version 2 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Library General Public License for more details.
#
# You should have received a copy of the GNU Library General Public
# License along with this library; if not, write to the 
# Free Software Foundation, Inc., 59 Temple Place - Suite 330, 
# Boston, MA  02111-1307  USA.

package Gnome2::GConf;

use 5.008;
use strict;
use warnings;

use Glib;

require Exporter;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use Gnome2::GConf ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);

our $VERSION = '1.043';

sub dl_load_flags { $^O eq 'darwin' ? 0x00 : 0x01 }

require XSLoader;
XSLoader::load('Gnome2::GConf', $VERSION);


# Preloaded methods go here

package Gnome2::GConf::Value;

use overload
	'==' => sub { Gnome2::GConf::Value::compare($_[0], $_[1]) },
	fallback => 1;

use overload
	'""' => sub { Gnome2::GConf::Value::to_string($_[0]) },
	fallback => 1;

package Gnome2::GConf::Client;
use Carp;

sub get_list
{
	my $self = shift;	# the object
	my $key  = shift;
	my $check_error = shift || 1;

	my $val = $self->get($key, $check_error);
	return $val->{value};
}

sub get_pair
{
	my $self = shift;	# the object
	my $key  = shift;
	my $check_error = shift || 1;

	my $val = $self->get($key, $check_error);
	carp "$key is not bound to a pair" if not $val->{type} eq 'pair';

	return ($val->{car}, $val->{cdr});
}

sub set_list
{
	my $self = shift;	# the object
	my $key  = shift;
	my $type = shift;
	my $list = shift;
	my $check_error = shift || 1;

	$self->set($key, { type => $type, value => $list }, $check_error);
}

sub set_pair
{
	my $self = shift;	# the object
	my $key  = shift;
	my $car  = shift;
	my $cdr  = shift;
	my $check_error = shift || 1;

	$self->set($key, {
			type	=> 'pair',
			car		=> $car,
			cdr		=> $cdr,
		}, $check_error);
}

package Gnome2::GConf;

1;
__END__

=head1 NAME

Gnome2::GConf - Perl wrappers for the GConf configuration engine.

=head1 SYNOPSIS

  use Gnome2::GConf;

  my $client = Gnome2::GConf::Client->get_default;
  my $app_key = "/apps/myapp/mykey";
  
  $client->add_dir($app_key, 'preload-none');
  
  # add a notify for the key
  my $notify_id = $client->notify_add($app_key, sub {
      my ($client, $cnxn_id, $entry) = @_;
      return unless $entry->{value};
		
      if ($entry->{value}->{type} eq 'string') {
        printf "key '%s' changed to '%s'\n",
	       $entry->{key},
	       $entry->{value}->{value};
      }
    });
  
  my $string = $client->get_string($app_key);
  $string = 'some string' unless $string;
  
  $client->set($app_key, { type => 'string', data => $string });
  
  # set a schema for the key
  $client->set_schema ($app_key,
      {
        type => 'string',
	locale => 'C',
	short_desc => 'Some key.',
	long_desc => 'This key does something.',
	owner => 'some_program'
      });
  
  # remove the notification callback
  $client->notify_remove($notify_id);

=head1 ABSTRACT

Perl bindings to the 2.2 series of the GConf configuration engine
libraries, for use with gtk2-perl.

=head1 DESCRIPTION

This module allows you to use the GConf configuration system in order
to store/retrieve the configuration of an application.  The GConf
system is a powerful configuration manager based on a user daemon that
handles a set of key and value pairs, and notifies any changes of the
value to every program that monitors those keys.  GConf is used
by GNOME 2.x.

To discuss gtk2-perl, ask questions and flame/praise the authors,
join gtk-perl-list@gnome.org at lists.gnome.org.

Find out more about Gnome at http://www.gnome.org.

=head1 DIFFERENT DATA TYPES

Some opaque data types in GConf are not registered inside the Glib type
system.  Thus, they have been implemented in a more perlish way, when
possible, for the sake of coherency and following the principle of
least surprise for the perl developer.  These changes try to preserve
semantics, to add syntactic sugar and to remove the need for accessor
methods.

=over

=item GConfEntry

See L<Gnome2::GConf::Entry>

=item GConfValue

See L<Gnome2::GConf::Value>

=item GConfChangeSet

See L<Gnome2::GConf::ChangeSet>

=item GConfSchema

See L<Gnome2::GConf::Schema>

=back

=head1 DIFFERENT CALL SIGNATURES

Reflecting the changes operated for the data types, some methods that use
those type have had the call signature modified.

=over

=item GConfNotifyFunc

In C, the function passed to C<Gnome2::GConf::notify_add> must have the
following signature:
	
	void (GConfNotifyFunc *) (GConfClient * client,
	                          guint cnxn_id,
	                          GConfEntry * entry);

Where C<GConfEntry> is a container for the key/value pair.  Since in
perl there's no C<GConfEntry> (see above), the C<entry> parameter is
an hashref.

=item GConfClient::get

=item GConfClient::set

In C, these accessor methods return/use a C<GConfValue>.  In perl, they
return/use an hashref.  See L<Gnome2::GConf::Value>

=item GConfClient::get_list

=item GConfClient::set_list

These accessor methods use a string for setting the type of the lists
(lists may have values of only B<one> type), and an arrayref containing
the values.

=item GConfClient::get_pair

=item GConfClient::set_pair

These accessor methods use two hashref (representing C<GConfValue>s)
for the C<car> and the C<cdr> parameters.

=item GConfClient::get_schema

=item GConfClient::set_schema

Similarly to the get/set pair above, these two methods return/use an
hashref. See L<Gnome2::GConf::Schema>.

=item GConfClient::commit_change_set

In C, this method return a boolean value (TRUE on success, FALSE on
failure). On user request (using the boolean parameter
C<remove_committed>), it also returns the C<GConfChangeSet>, pruned of
the successfully committed keys.  In perl, this method returns a
boolean value both in scalar context or if the user sets to FALSE the
C<remove_committed> parameter; in array context or if the user requests
the uncommitted keys, returns both the return value and the pruned
C<GConfChangeSet>.

=back

=head1 SEE ALSO

L<perl>(1), L<Glib>(3pm).

=head1 AUTHOR

Emmanuele Bassi E<lt>ebassi@gmail.comE<gt>

gtk2-perl created by the gtk2-perl team: http://gtk2-perl.sf.net

=head1 COPYRIGHT AND LICENSE

Copyright 2003-2006 by Emmanuele Bassi

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Library General Public
License as published by the Free Software Foundation; either
version 2 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Library General Public License for more details.

You should have received a copy of the GNU Library General Public
License along with this library; if not, write to the 
Free Software Foundation, Inc., 59 Temple Place - Suite 330, 
Boston, MA  02111-1307  USA.

=cut


gconfperl.h
/*
 * Copyright (c) 2003-2005 by Emmanuele Bassi (see the file AUTHORS)
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the 
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330, 
 * Boston, MA  02111-1307  USA.
 */

#ifndef _GNOME_GCONF_PERL_H_
# define _GNOME_GCONF_PERL_H_

# include "gperl.h"

/* basic include files */
# include <gconf/gconf-changeset.h>
# include <gconf/gconf-client.h>
# include <gconf/gconf-engine.h>
# include <gconf/gconf-enum-types.h>
# include <gconf/gconf-error.h>
# include <gconf/gconf-listeners.h>
# include <gconf/gconf-schema.h>
# include <gconf/gconf-value.h>
# include <gconf/gconf.h>

#ifndef GCONFPERL_TYPE_GCONF_ERROR
# define GCONFPERL_TYPE_GCONF_ERROR (gconfperl_gconf_error_get_type ())
GType gconfperl_gconf_error_get_type (void) G_GNUC_CONST;
#endif

#ifndef GCONF_TYPE_ENGINE
# define GCONF_TYPE_ENGINE (gconfperl_gconf_engine_get_type ())
GType gconfperl_gconf_engine_get_type (void) G_GNUC_CONST;
#endif

/* forward declaration for opaque containers converters */
SV * newSVGConfEntry (GConfEntry *);
SV * newSVGConfValue (GConfValue *);
SV * newSVGConfSchema (GConfSchema *);
SV * newSVGConfChangeSet (GConfChangeSet *);

GConfEntry * SvGConfEntry (SV *);
GConfValue * SvGConfValue (SV *);
GConfSchema * SvGConfSchema (SV *);
GConfChangeSet * SvGConfChangeSet (SV *);

# include "gconfperl-autogen.h"
# include "gconfperl-version.h"

#endif /* _GNOME_GCONF_PERL_H_ */



Makefile.PL
#
# $Header: /cvsroot/gtk2-perl/gtk2-perl-xs/Gnome2-GConf/Makefile.PL,v 1.19 2006/06/20 09:15:09 ebassi Exp $
#

use 5.008;
use strict;
use warnings;
use Cwd;
use File::Spec;
use ExtUtils::MakeMaker;

# minimum required version of dependancies we need to build
our %build_reqs = (
	'perl-ExtUtils-Depends'   => '0.2',
	'perl-ExtUtils-PkgConfig' => '1.03',
	'perl-Glib'               => '1.120', # for Glib::MakeHelper
	'GConf'                   => '2.0.0',
);

# Writing a fake Makefile ensures that CPAN will pick up the correct
# dependencies and install them.
unless (eval "use ExtUtils::Depends '$build_reqs{'perl-ExtUtils-Depends'}';"
           . "use ExtUtils::PkgConfig '$build_reqs{'perl-ExtUtils-PkgConfig'}';"
	   . "use Glib::CodeGen;"
           # just seeing if Glib is available isn't enough, make sure
           # it's recent enough, too 
           . "use Glib '$build_reqs{'perl-Glib'}';"
           . "use Glib::MakeHelper;"
           . "1") {
   warn "$@\n";
   WriteMakefile(
         PREREQ_FATAL => 1,
         PREREQ_PM    => {
            'Glib'                => $build_reqs{'perl-Glib'},
            'ExtUtils::Depends'   => $build_reqs{'perl-ExtUtils-Depends'},
            'ExtUtils::PkgConfig' => $build_reqs{'perl-ExtUtils-PkgConfig'},
         },
   );
   exit 1; # not reached
}

mkdir 'build', 0777;

our %pkgcfg = ExtUtils::PkgConfig->find ('gconf-2.0 >= '.$build_reqs{GConf});


our @xs_files = <xs/*.xs>;
our %pm_files = (
    'GConf.pm' => '$(INST_LIBDIR)/GConf.pm',
);
our %pod_files = (
    'GConf.pm' => '$(INST_MAN3DIR)/Gnome2::GConf.$(MAN3EXT)',
    Glib::MakeHelper->do_pod_files (@xs_files),
);

ExtUtils::PkgConfig->write_version_macros ("build/gconfperl-version.h",
                                           'gconf-2.0' => 'GCONF');

#
# autogeneration
#
Glib::CodeGen->parse_maps ('gconfperl');
Glib::CodeGen->write_boot (ignore => '^Gnome2::GConf$');

# now we're ready to start creating the makefile.
# we need to use ExtUtils::Depends to get relevant information out of
# the Glib extension, and to save config information for other modules which
# will chain from this one.
our $gconf = ExtUtils::Depends->new ('Gnome2::GConf', 'Glib');
$gconf->set_inc ($pkgcfg{cflags} . ' -I./build ');
$gconf->set_libs ($pkgcfg{libs});
$gconf->add_xs (@xs_files);
$gconf->add_pm (%pm_files);
my $cwd = cwd();
$gconf->add_typemaps (map { File::Spec->catfile($cwd, $_) }
                      'build/gconfperl.typemap', 'gconf.typemap');
$gconf->install (qw(gconfperl.h
		build/gconfperl-autogen.h
		build/gconfperl-version.h));
$gconf->save_config ('build/IFiles.pm');

WriteMakefile(
    NAME          => 'Gnome2::GConf',
    VERSION_FROM  => 'GConf.pm', # finds $VERSION
    PREREQ_PM     => {}, # e.g., Module::Name => 1.1
    ABSTRACT_FROM => 'GConf.pm', # retrieve abstract from module
    XSPROTOARG    => '-noprototypes',
    MAN3PODS	  => \%pod_files,
    $gconf->get_makefile_vars,
);

#print "
#WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING
#
#This is an unstable development release of Gnome2::GConf.  The API is
#not frozen and things are subject to change at any time.  Report any
#bugs to gtk-perl-list AT gnome DOT org as soon as possible.
#Please use the stable 1.020 version for important work.
#
#WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING
#
#";
			

sub MY::postamble
{
	return Glib::MakeHelper->postamble_clean ()
	     . Glib::MakeHelper->postamble_docs_full (
		DEPENDS => $gconf,
		DOCTYPES => 'doctypes',
		COPYRIGHT_FROM => 'copyright.pod',
	       )
	     . Glib::MakeHelper->postamble_rpms (
		'GCONF' => $build_reqs{'GConf'},
		'PERL_EXTUTILS_DEPENDS' => 
			$build_reqs{'perl-ExtUtils-Depends'},
		'PERL_EXTUTILS_PKGCONFIG' => 
			$build_reqs{'perl-ExtUtils-PkgConfig'},
		'PERL_GLIB' => $build_reqs{'perl-Glib'},
	       );
}
